pkg/kubecost/allocation.go

```diff
@@ -2261,7 +2261,71 @@ func (asr *AllocationSetRange) Accumulate() (*AllocationSet, error) {
 }
 
 // TODO accumulate into lower-resolution chunks of the given resolution
-// func (asr *AllocationSetRange) AccumulateBy(resolution time.Duration) *AllocationSetRange
+func (asr *AllocationSetRange) AccumulateBy(resolution time.Duration) (*AllocationSetRange, error) {
+	allocSetRange := &AllocationSetRange{allocations: []*AllocationSet{}}
+	var allocSet *AllocationSet
+	var err error
+
+	asr.Lock()
+	defer asr.Unlock()
+
+	// use window() with no lock
+	if asr.window().IsEmpty() {
+		return asr, nil
+	}
+
+	// Call total accumulate func if resolution is greater than total window duration
+	if resolution > asr.window().Duration() {
+		// unable to acquire lock here if old asr.accumulate is called
+		for _, as := range asr.allocations {
+			allocSet, err = allocSet.accumulate(as)
+			if err != nil {
+				return nil, err
+			}
+		}
+		allocSetRange.allocations = append(allocSetRange.allocations, allocSet)
+		return allocSetRange, nil
+	}
+
+	// two ways to get end of asr window
+	// 1. check if last set of asr
+	// 2. check if set window.end is asr.window.end
+	// both require no lock to be present
+	// option 1: 2 asr.window methods
+	// option 2: 2 window.length methods
+	asrWindow := asr.window()
+	asrResolution := asrWindow.Duration()
+
+	// if asrResolution < 1 hour then there is not enough data to accumulate
+	// if as.allocations(end - start) > resolution then no need, it's already accumulated. Requery to make more granular in the future?
+	// check this again?
+	if asrResolution <= time.Hour || asrResolution < resolution {
+		return asr, nil
+	}
+
+	// do not compound
+	// check as.window and accumulate till windowSum == time.duration wanted -> add accumulated set to allocSetRange
+	var currAccumulatedSum time.Duration
+	for _, as := range asr.allocations {
+		allocSet, err = allocSet.accumulate(as)
+		if err != nil {
+			return nil, err
+		}
+
+		if allocSet != nil {
+			currAccumulatedSum += allocSet.Window.Duration()
+
+			// check if end of asr to sum the final set too
+			if currAccumulatedSum >= resolution || NewWindow(nil, asrWindow.end).Equal(NewWindow(nil, allocSet.Window.end)) {
+				allocSetRange.allocations = append(allocSetRange.allocations, allocSet)
+				allocSet = &AllocationSet{allocations: map[string]*Allocation{}}
+				currAccumulatedSum = 0
+			}
+		}
+	}
```

Comment:
If you want to know if you have the final `AllocationSet` in `asr.allocations`, you can actually change your loop to include the index in the slice:

```go
for i, as := range asr.allocations {
    // ...

    if i == len(asr.allocations) - 1 {
        // do something because this is the last AllocationSet in asr
    } else if currAccumulatedSum >= resolution {
        // do something
    }
}
```